### PR TITLE
Ports TG hit_reaction()

### PR DESCRIPTION
**Pull request by Aurorablade**
ports https://github.com/tgstation/tgstation/pull/13327

Pretty much makes what happens when you are hit more dynamic per item.

I didn't test SHIT yet and i probaly have merge conflicts already...I am tired and need to reset up my local server config and get the files set up.

I also probaly missed some things i hope someone points out to me. I am tired.

:cl: Fethas and TG
rscdel:Removes IsShield
rscadd:Adds hit reactions to replace Isheilds
rscadd:Hitting someone dressed like a clown now makes them honk.
/:cl:

### CODE REVIEW

**Comment by Spacemanspark** [comment]
Two people going at it with Eswords should have some awesome effects.

**Comment by FalseIncarnate** [comment]
Want to point out that you renamed the crate and objectives for Ablative Armor vests, but didn't rename the armor itself.

File: code/game/machinery/computer/HolodeckControl.dm [hunk 1]
```diff
@@ -432,8 +432,8 @@
 	icon_state = "claymorered"
 	item_state = "claymorered"
 
-/obj/item/weapon/holo/claymore/IsShield()
-	return 1
+/obj/item/weapon/holo/claymore/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance)
+	return ..()
```
Comment (Fox-McCloud) [inline code/game/machinery/computer/HolodeckControl.dm]: don't you need to specificy `block_chance`?

File: code/game/objects/items/weapons/holy_weapons.dm [hunk 2]
```diff
@@ -77,8 +77,11 @@
 	force = 5
 	slot_flags = SLOT_BACK
 
-/obj/item/weapon/nullrod/staff/IsShield()
-	return 1
+/obj/item/weapon/nullrod/claymore/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance, damage, attack_type)
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/holy_weapons.dm]: errrrr? Wrong thing copied over?

File: code/game/objects/items/weapons/holy_weapons.dm [hunk 3]
```diff
@@ -342,11 +344,10 @@
 	icon_state = "bostaff0"
 	item_state = "bostaff0"
 
-/obj/item/weapon/nullrod/claymore/bostaff/IsShield()
-	if(prob(40))
-		return 1
-	else
-		return 0
+/obj/item/weapon/nullrod/claymore/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance, damage, attack_type)
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/holy_weapons.dm]: errrrr? Wrong thing copied over?

File: code/game/objects/items/weapons/shields.dm [hunk 4]
```diff
@@ -42,11 +46,10 @@
 	item_state = "buckler"
 	materials = list()
 
-/obj/item/weapon/shield/riot/buckler/IsShield()
+/obj/item/weapon/shield/riot/buckler/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance)
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/shields.dm]: block_chance instead of probabilities

File: code/modules/clothing/suits/armor.dm [hunk 5]
```diff
@@ -250,6 +246,32 @@
 	src.item_state = "reactiveoff"
 	..()
 
+/obj/item/clothing/suit/armor/reactive/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance)
```
Comment (Fox-McCloud) [inline code/modules/clothing/suits/armor.dm]: I'd use TG's more recent version of this minus the radiation and exempting space turfs.

No one uses the reactive armor on our codebase.

File: code/modules/clothing/under/jobs/civilian.dm [hunk 6]
```diff
@@ -57,6 +57,10 @@
 	item_color = "clown"
 	flags = ONESIZEFITSALL
 
+/obj/item/clothing/under/rank/clown/hit_reaction()
+	playsound(loc, 'sound/items/bikehorn.ogg', 50, 1, -1)
```
Comment (Fox-McCloud) [inline code/modules/clothing/under/jobs/civilian.dm]: YESSSSSS

**Comment by Fox-McCloud** [comment]
You'll want to remove the `NO_SHIELD` flag as well, in this PR.

The armor refactor already put armor penetration on everything that needs it (bullets and melee), with this port, NO_SHIELD will be obsolete.

**Comment by Ar3nn** [comment]
She's directly grabbed their hit_reaction feature PR. It's labeled a fix on their end, but the PR she linked is their refactor. Could probably go suggest a few additional PRs for her to look at to do some of the stuff you want it to do?

File: code/game/machinery/computer/HolodeckControl.dm (hunk 1 above)
Comment (Aurorablade) [inline code/game/machinery/computer/HolodeckControl.dm]: i wasn't sure about holoweapons at the time...

File: code/game/objects/items/weapons/holy_weapons.dm (hunk 2 above)
Comment (Aurorablade) [inline code/game/objects/items/weapons/holy_weapons.dm]: *flails at fox*

File: code/game/objects/items/weapons/holy_weapons.dm (hunk 3 above)
Comment (Aurorablade) [inline code/game/objects/items/weapons/holy_weapons.dm]: *FLAILS AT FOX*

**Comment by Aurorablade** [comment]
i didn't link all the PRs i looked through..and yes i porobaly missed something.

File: code/game/objects/items/weapons/shields.dm [hunk 7]
```diff
@@ -45,11 +45,10 @@
 	icon_state = "buckler"
 	item_state = "buckler"
 	materials = list()
+	block_chance = 60
 
 /obj/item/weapon/shield/riot/buckler/hit_reaction(mob/living/carbon/human/owner, attack_text, final_block_chance)
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/shields.dm]: this proc doesn't need to exist

File: code/__DEFINES/flags.dm [hunk 8]
```diff
@@ -4,11 +4,10 @@
...
 #define HANDSLOW        		16		// If an item has this flag, it will slow you to carry it
-#define NOSHIELD				32		// weapon not affected by shield
-#define CONDUCT					64		// conducts electricity (metal etc.)
-#define ABSTRACT				128		// for all things that are technically items but used for various different stuff, made it 128 because it could conflict with other flags other way
-#define ON_BORDER				512		// item has priority to check when entering or leaving
-#define NODELAY 				32768	// 1 second attackby delay skipped (Can be used once every 0.2s). Most objects have a 1s attackby delay, which doesn't require a flag.
+#define CONDUCT					32		// conducts electricity (metal etc.)
+#define ABSTRACT				64		// for all things that are technically items but used for various different stuff, made it 128 because it could conflict with other flags other way
+#define ON_BORDER				128		// item has priority to check when entering or leaving
+#define NODELAY 				512		// 1 second attackby delay skipped (Can be used once every 0.2s). Most objects have a 1s attackby delay, which doesn't require a flag.
```
Comment (Fox-McCloud) [inline code/__DEFINES/flags.dm]: More flags is always a good thing. Huzzah.

File: code/game/objects/items/weapons/staff.dm [hunk 9]
```diff
@@ -8,8 +8,11 @@
 	throw_speed = 1
 	throw_range = 5
 	w_class = 2.0
+<<<<<<< HEAD
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/staff.dm]: nice merge conflicts you got dere

File: code/game/objects/items/weapons/twohanded.dm [hunk 10]
```diff
@@ -205,13 +197,17 @@
 	force_wielded = 34
 	wieldsound = 'sound/weapons/saberon.ogg'
 	unwieldsound = 'sound/weapons/saberoff.ogg'
+<<<<<<< HEAD
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/twohanded.dm]: :eyes:

File: code/game/objects/items/weapons/twohanded.dm [hunk 11]
```diff
@@ -371,8 +366,11 @@
 	force_wielded = 40  //you'll gouge their eye out! Or a limb...maybe even their entire body!
 	wieldsound = 'sound/weapons/chainsawstart.ogg'
 	hitsound = null
+<<<<<<< HEAD
```
Comment (tigercat2000) [inline code/game/objects/items/weapons/twohanded.dm]: feethass pls

**Comment by TheDZD** [comment]
[Well, there's your problem.](https://github.com/Aurorablade/Paradise/blob/HITME/code/game/objects/items/weapons/staff.dm#L76)

File: code/game/objects/items/weapons/shields.dm [hunk 12]
```diff
@@ -41,12 +45,7 @@
 	icon_state = "buckler"
 	item_state = "buckler"
 	materials = list()
-
-/obj/item/weapon/shield/riot/buckler/IsShield()
-	if(prob(60))
-		return 1
-	else
-		return 0
+	block_chance = 60
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/shields.dm]: I believe this is meant to have a 30% chance, not 60%-it's worse than a riot shield, not better.

File: code/modules/clothing/suits/armor.dm [hunk 13]
```diff
@@ -213,41 +213,76 @@
...
+	if(!active)
+		return 0
+	if(prob(hit_reaction_chance))
+		var/mob/living/carbon/human/H = owner
+		if(world.time < reactivearmor_cooldown)
+			owner.visible_message("<span class='danger'>The reactive teleport system is still recharging! It fails to teleport [H]!</span>")
+			return
+		owner.visible_message("<span class='danger'>The reactive teleport system flings [H] clear of [attack_text], shutting itself off in the process!</span>")
+		var/list/turfs = new/list()
+		for(var/turf/T in orange(tele_range, H))
```
Comment (Fox-McCloud) [inline code/modules/clothing/suits/armor.dm]: Still think this should exempt space and not have the cooldown it currently does---it should also unbuckle the mob first (like it currently does).

Reason being is..well, reactive tele-armor is shunned, currently. Though all these various tweaks could easily be put in place if it starts becoming problematic. At least, IMO.

**Comment by Fox-McCloud** [comment]
@Aurorablade has asked me to take this on due to issues she's encountered; I will do so.

**Pull request closed without merging**